### DISCUSSION TITLE
How would I implement a virtual filesystem in Deno?

### DISCUSSION THREAD

**Discussion by tmcw** [denoland/deno#20151] [General]
Deno's permissions model has been great so far, but supporting existing npm & Deno modules can be tough when the setting is to entirely block filesystem & environment access. It'd be great to support something like Deno Deploy's [virtual filesystem](https://deno.com/subhosting/docs/virtual-filesystem) to essentially run Deno but use custom implementations of the ext/fs crate.

Is there a path to doing this - using Deno as a crate and swapping a dependency, or is it better to try and do this as a sort of 'operating system' level task? Thanks!

**Comment by bartlomieju** [comment]
If you need to swap out an implementation of `ext/fs` you will pretty much have to implement the layer as you need it and compile Deno yourself to produce the final binary. We don't provide an implementation of VFS that can be easily enabled or swapped out. If that's not an option to compile it yourself then probably going with "operating system level" approach would be better.
